File: app/src/main/java/com/pastforward/data/network/GeminiService.kt
package com.pastforward.data.network

import android.util.Log
import com.google.gson.Gson
import com.pastforward.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit
import kotlin.math.pow

object GeminiService {
    private const val TAG = "GeminiService"
    private val client = OkHttpClient.Builder()
        .connectTimeout(60, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()
    
    private val gson = Gson()
    private val mediaTypeJson = "application/json; charset=utf-8".toMediaType()

    // Endpoint for gemini-2.5-flash-image
    private const val MODEL_NAME = "gemini-2.5-flash-image"
    private const val BASE_URL = "https://generativelanguage.googleapis.com/v1beta/models/$MODEL_NAME:generateContent"

    // Fallback prompt template
    fun getFallbackPrompt(decadeOrWorld: String): String {
        return "Create a photograph of the person in this image as if they were living in the $decadeOrWorld. " +
                "The photograph should capture the distinct fashion, hairstyles, and overall atmosphere of that time period or environment. " +
                "Ensure the final image is a clear photograph that looks authentic."
    }

    suspend fun generateStyledImage(
        base64Image: String,
        prompt: String,
        fallbackStyle: String? = null
    ): Result<String> = withContext(Dispatchers.IO) {
        val apiKey = BuildConfig.GEMINI_API_KEY
        if (apiKey.isEmpty()) {
            return@withContext Result.failure(Exception("Gemini API key is not configured. Please check your environment variables."))
        }

        // Clean base64 if it has header like "data:image/jpeg;base64,"
        val cleanBase64 = if (base64Image.contains(",")) {
            base64Image.substringAfter(",")
        } else {
            base64Image
        }

        // Detect mimeType (default to image/jpeg)
        val mimeType = if (base64Image.contains("image/png")) "image/png" else "image/jpeg"

        val url = "$BASE_URL?key=$apiKey"

        Log.d(TAG, "Starting generation with prompt: $prompt")
        var lastError: Throwable? = null

        // Attempt 1: Call Gemini with Retry and the original prompt
        try {
            val base64Response = callGeminiWithRetry(cleanBase64, mimeType, prompt, url)
            return@withContext Result.success(base64Response)
        } catch (e: Exception) {
            Log.w(TAG, "Original prompt failed, checking if it was blocked or rate limited. Error: ${e.message}")
            lastError = e
        }

        // Attempt 2: Fallback prompt if fallbackStyle is provided and the original is censored/failed
        if (fallbackStyle != null) {
            try {
                val fallbackPrompt = getFallbackPrompt(fallbackStyle)
                Log.d(TAG, "Attempting generation with fallback prompt: $fallbackPrompt")
                val base64Response = callGeminiWithRetry(cleanBase64, mimeType, fallbackPrompt, url)
                return@withContext Result.success(base64Response)
            } catch (e: Exception) {
                Log.e(TAG, "Fallback prompt also failed: ${e.message}")
                lastError = e
            }
        }

        return@withContext Result.failure(lastError ?: Exception("Unknown image generation error"))
    }

    private suspend fun callGeminiWithRetry(
        base64Image: String,
        mimeType: String,
        prompt: String,
        url: String
    ): String {
        val maxRetries = 4
        val initialDelay = 2000L

        for (attempt in 1..maxRetries) {
            try {
                return executeApiRequest(base64Image, mimeType, prompt, url)
            } catch (e: Exception) {
                Log.e(TAG, "API attempt $attempt/$maxRetries failed: ${e.message}")
                val errorMsg = e.message ?: ""
                val isRateLimit = errorMsg.contains("429") || errorMsg.contains("quota") || errorMsg.contains("RESOURCE_EXHAUSTED")
                val isInternal = errorMsg.contains("500") || errorMsg.contains("INTERNAL") || errorMsg.contains("overloaded")

                if ((isRateLimit || isInternal) && attempt < maxRetries) {
                    val baseDelay = if (isRateLimit) 8000L else initialDelay
                    // Exponential backoff + jitter
                    val delay = (baseDelay * 2.0.pow(attempt - 1).toLong()) + (0..1000).random()
                    Log.d(TAG, "Retrying in ${delay}ms. Reason: ${if (isRateLimit) "Rate Limit" else "Internal Error"}")
                    kotlinx.coroutines.delay(delay)
                    continue
                }

                // Handle friendly errors
                if (isRateLimit) {
                    if (errorMsg.contains("quota")) {
                        throw Exception("Quota exceeded. Please check your Gemini API billing or wait for the quota to reset.")
                    }
                    throw Exception("Rate limit exceeded. The AI is busy, please wait a minute and try again.")
                }
                throw e
            }
        }
        throw Exception("API call failed after max retries.")
    }

    private fun executeApiRequest(
        base64Image: String,
        mimeType: String,
        prompt: String,
        url: String
    ): String {
        // Construct standard REST JSON payload equivalent to Javascript client
        val payload = mapOf(
            "contents" to listOf(
                mapOf(
                    "parts" to listOf(
                        mapOf(
                            "inlineData" to mapOf(
                                "mimeType" to mimeType,
                                "data" to base64Image
                            )
                        ),
                        mapOf(
                            "text" to prompt
                        )
                    )
                )
            )
        )

        val jsonString = gson.toJson(payload)
        val body = jsonString.toRequestBody(mediaTypeJson)
        val request = Request.Builder()
            .url(url)
            .post(body)
            .build()

        client.newCall(request).execute().use { response ->
            val responseBody = response.body?.string() ?: ""
            if (!response.isSuccessful) {
                throw IOException("Gemini API Error (HTTP ${response.code}): $responseBody")
            }

            // Parse response
            val parsed = gson.fromJson(responseBody, GeminiResponse::class.java)
            val inlineData = parsed.candidates?.firstOrNull()
                ?.content?.parts?.firstOrNull()
                ?.inlineData

            if (inlineData != null) {
                return "data:${inlineData.mimeType};base64,${inlineData.data}"
            }

            // If response text is present but no image, check safety block text
            val fallbackText = parsed.candidates?.firstOrNull()?.content?.parts?.firstOrNull()?.text ?: ""
            if (fallbackText.lowercase().contains("safety") || fallbackText.isEmpty()) {
                throw SafetyException("Content filtered for safety. Try a different photo.")
            }

            throw IOException("Gemini API replied with text instead of image: \"$fallbackText\"")
        }
    }
}

// Custom safety block exception
class SafetyException(message: String) : IOException(message)

// GSON classes mapping standard Gemini API Response structure
data class GeminiResponse(val candidates: List<GeminiCandidate>?)
data class GeminiCandidate(val content: GeminiContent?)
data class GeminiContent(val parts: List<GeminiPart>?)
data class GeminiPart(val text: String?, val inlineData: GeminiInlineData?)
data class GeminiInlineData(val mimeType: String, val data: String)

File: services/geminiService.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
*/
import { GoogleGenAI } from "@google/genai";
import type { GenerateContentResponse } from "@google/genai";

let aiInstance: GoogleGenAI | null = null;

function getAI() {
    if (!aiInstance) {
        const apiKey = process.env.API_KEY || process.env.GEMINI_API_KEY;
        if (!apiKey) {
            throw new Error("Gemini API key is not configured. Please set GEMINI_API_KEY in your environment.");
        }
        aiInstance = new GoogleGenAI({ apiKey });
    }
    return aiInstance;
}

// --- Helper Functions ---

/**
 * Creates a fallback prompt to use when the primary one is blocked.
 * @param decade The decade string (e.g., "1950s").
 * @returns The fallback prompt string.
 */
function getFallbackPrompt(decade: string): string {
    return `Create a photograph of the person in this image as if they were living in the ${decade}. The photograph should capture the distinct fashion, hairstyles, and overall atmosphere of that time period. Ensure the final image is a clear photograph that looks authentic to the era.`;
}

/**
 * Extracts the style (e.g., "1950s" or "Cyberpunk") from a prompt string.
 * @param prompt The original prompt.
 * @returns The style string or null if not found.
 */
function extractStyle(prompt: string): string | null {
    // Match decades (e.g., "1950s") or specific keywords from the prompt
    const match = prompt.match(/style of the (.*?)\./);
    return match ? match[1] : null;
}

/**
 * Processes the Gemini API response, extracting the image or throwing an error if none is found.
 * @param response The response from the generateContent call.
 * @returns A data URL string for the generated image.
 */
function processGeminiResponse(response: GenerateContentResponse): string {
    const imagePartFromResponse = response.candidates?.[0]?.content?.parts?.find(part => part.inlineData);

    if (imagePartFromResponse?.inlineData) {
        const { mimeType, data } = imagePartFromResponse.inlineData;
        return `data:${mimeType};base64,${data}`;
    }

    const textResponse = response.text;
    
    // Check for safety filters or other common reasons for no image
    if (textResponse?.toLowerCase().includes("safety") || !textResponse) {
        throw new Error("Content filtered for safety. Try a different photo.");
    }

    console.error("API did not return an image. Response:", textResponse);
    throw new Error("The AI model couldn't generate an image for this style.");
}

/**
 * A wrapper for the Gemini API call that includes a retry mechanism for internal server errors.
 * @param imagePart The image part of the request payload.
 * @param textPart The text part of the request payload.
 * @returns The GenerateContentResponse from the API.
 */
async function callGeminiWithRetry(imagePart: object, textPart: object): Promise<GenerateContentResponse> {
    const maxRetries = 4;
    const initialDelay = 2000;

    for (let attempt = 1; attempt <= maxRetries; attempt++) {
        try {
            const ai = getAI();
            return await ai.models.generateContent({
                model: 'gemini-2.5-flash-image',
                contents: { parts: [imagePart, textPart] },
            });
        } catch (error) {
            console.error(`Error calling Gemini API (Attempt ${attempt}/${maxRetries}):`, error);
            const errorMessage = error instanceof Error ? error.message : JSON.stringify(error);
            
            const isRateLimit = errorMessage.includes('429') || errorMessage.includes('quota') || errorMessage.includes('RESOURCE_EXHAUSTED');
            const isInternalError = errorMessage.includes('"code":500') || errorMessage.includes('INTERNAL') || errorMessage.includes('overloaded');

            if ((isInternalError || isRateLimit) && attempt < maxRetries) {
                // Significantly longer delay for rate limits to allow quota to reset
                const baseDelay = isRateLimit ? 10000 : initialDelay;
                const delay = baseDelay * Math.pow(2, attempt - 1) + (Math.random() * 2000);
                
                console.log(`${isRateLimit ? 'Rate limit' : 'Internal error'} detected. Retrying in ${Math.round(delay)}ms...`);
                await new Promise(resolve => setTimeout(resolve, delay));
                continue;
            }
            
            if (isRateLimit) {
                if (errorMessage.includes('quota')) {
                    throw new Error("Quota exceeded. Please check your Gemini API billing or wait for the quota to reset.");
                }
                throw new Error("Rate limit exceeded. The AI is busy, please try again in a minute.");
            }

            throw error;
        }
    }
    // This should be unreachable due to the loop and throw logic above.
    throw new Error("Gemini API call failed after all retries.");
}


/**
 * Generates a styled image from a source image and a prompt.
 * It includes a fallback mechanism for prompts that might be blocked in certain regions.
 * @param imageDataUrl A data URL string of the source image (e.g., 'data:image/png;base64,...').
 * @param prompt The prompt to guide the image generation.
 * @returns A promise that resolves to a base64-encoded image data URL of the generated image.
 */
export async function generateStyledImage(imageDataUrl: string, prompt: string): Promise<string> {
  const match = imageDataUrl.match(/^data:(image\/\w+);base64,(.*)$/);
  if (!match) {
    throw new Error("Invalid image data URL format. Expected 'data:image/...;base64,...'");
  }
  const [, mimeType, base64Data] = match;

    const imagePart = {
        inlineData: { mimeType, data: base64Data },
    };

    // --- First attempt with the original prompt ---
    try {
        console.log("Attempting generation with original prompt...");
        const textPart = { text: prompt };
        const response = await callGeminiWithRetry(imagePart, textPart);
        return processGeminiResponse(response);
    } catch (error) {
        const errorMessage = error instanceof Error ? error.message : JSON.stringify(error);
        const isNoImageError = errorMessage.includes("The AI model responded with text instead of an image");

        if (isNoImageError) {
            console.warn("Original prompt was likely blocked. Trying a fallback prompt.");
            const style = extractStyle(prompt);
            if (!style) {
                console.error("Could not extract style from prompt, cannot use fallback.");
                throw error; // Re-throw the original "no image" error.
            }

            // --- Second attempt with the fallback prompt ---
            try {
                const fallbackPrompt = getFallbackPrompt(style);
                console.log(`Attempting generation with fallback prompt for ${style}...`);
                const fallbackTextPart = { text: fallbackPrompt };
                const fallbackResponse = await callGeminiWithRetry(imagePart, fallbackTextPart);
                return processGeminiResponse(fallbackResponse);
            } catch (fallbackError) {
                console.error("Fallback prompt also failed.", fallbackError);
                const finalErrorMessage = fallbackError instanceof Error ? fallbackError.message : String(fallbackError);
                throw new Error(`The AI model failed with both original and fallback prompts. Last error: ${finalErrorMessage}`);
            }
        } else {
            // This is for other errors, like a final internal server error after retries.
            console.error("An unrecoverable error occurred during image generation.", error);
            const userFriendlyMessage = errorMessage.includes("limit") ? "Rate limit reached. Try again in a minute." : "Server error. Please try again.";
            throw new Error(userFriendlyMessage);
        }
    }
}
